### PR TITLE
Propagate grants to compressed hypertable

### LINKED ISSUE #3209: Grants do not propagate to compressed hypertable
**Relevant system information:**
 - OS: Ubuntu 20.04
 - PostgreSQL version (output of `postgres --version`): All
 - TimescaleDB version (output of `\dx` in `psql`): 2.2
 - Installation method: Source

**Describe the bug**

If a user is granted permissions on a hypertable and compression is enabled, the compressed hypertable does not get grants.

**To Reproduce**

```
postgres=# CREATE TABLE conditions(
postgres(#     time TIMESTAMPTZ NOT NULL,
postgres(#     device INTEGER,
postgres(#     temperature FLOAT
postgres(# );
CREATE TABLE
postgres=# SELECT * FROM create_hypertable('conditions', 'time',
postgres(#        chunk_time_interval => INTERVAL '1 day');
 hypertable_id | schema_name | table_name | created 
---------------+-------------+------------+---------
             3 | public      | conditions | t
(1 row)

postgres=# 
postgres=# INSERT INTO conditions
postgres-# SELECT time, (random()*30)::int, random()*80 - 40
postgres-# FROM generate_series(NOW() - INTERVAL '90 days', NOW(), '1 min') AS time;
INSERT 0 129541
postgres=# 
postgres=# ALTER TABLE conditions SET (
postgres(#     timescaledb.compress,
postgres(#     timescaledb.compress_segmentby = 'device'
postgres(# );
ALTER TABLE
postgres=# \dp conditions
                               Access privileges
 Schema |    Name    | Type  | Access privileges | Column privileges | Policies 
--------+------------+-------+-------------------+-------------------+----------
 public | conditions | table |                   |                   | 
(1 row)

postgres=# SELECT compress_chunk(show_chunks('conditions'));
              compress_chunk              
------------------------------------------
    .
    .
    .
 _timescaledb_internal._hyper_3_273_chunk
(91 rows)

postgres=# select * from _timescaledb_catalog.hypertable;                 
-[ RECORD 1 ]------------+-------------------------
id                       | 4
schema_name              | _timescaledb_internal
table_name               | _compressed_hypertable_4
associated_schema_name   | _timescaledb_internal
associated_table_prefix  | _hyper_4
num_dimensions           | 0
chunk_sizing_func_schema | _timescaledb_internal
chunk_sizing_func_name   | calculate_chunk_interval
chunk_target_size        | 0
compression_state        | 2
compressed_hypertable_id | [NULL]
replication_factor       | [NULL]
-[ RECORD 2 ]------------+-------------------------
id                       | 3
schema_name              | public
table_name               | conditions
associated_schema_name   | _timescaledb_internal
associated_table_prefix  | _hyper_3
num_dimensions           | 1
chunk_sizing_func_schema | _timescaledb_internal
chunk_sizing_func_name   | calculate_chunk_interval
chunk_target_size        | 0
compression_state        | 1
compressed_hypertable_id | 4
replication_factor       | [NULL]

postgres=# \dp _timescaledb_internal._compressed_hypertable_4 
                                              Access privileges
        Schema         |           Name           | Type  | Access privileges | Column privileges | Policies 
-----------------------+--------------------------+-------+-------------------+-------------------+----------
 _timescaledb_internal | _compressed_hypertable_4 | table |                   |                   | 
(1 row)

postgres=# \dp conditions
                               Access privileges
 Schema |    Name    | Type  | Access privileges | Column privileges | Policies 
--------+------------+-------+-------------------+-------------------+----------
 public | conditions | table |                   |                   | 
(1 row)

postgres=# \du
                                   List of roles
 Role name |                         Attributes                         | Member of 
-----------+------------------------------------------------------------+-----------
 admin     | Cannot login                                               | {}
 postgres  | Superuser, Create role, Create DB, Replication, Bypass RLS | {}
 tsdbadmin | Cannot login                                               | {}

postgres=# grant all on conditions to tsdbadmin ;
GRANT
postgres=# \dp conditions
                                    Access privileges
 Schema |    Name    | Type  |     Access privileges      | Column privileges | Policies 
--------+------------+-------+----------------------------+-------------------+----------
 public | conditions | table | postgres=arwdDxt/postgres +|                   | 
        |            |       | tsdbadmin=arwdDxt/postgres |                   | 
(1 row)

postgres=# \dp _timescaledb_internal._compressed_hypertable_4 
                                              Access privileges
        Schema         |           Name           | Type  | Access privileges | Column privileges | Policies 
-----------------------+--------------------------+-------+-------------------+-------------------+----------
 _timescaledb_internal | _compressed_hypertable_4 | table |                   |                   | 
(1 row)
```

**Expected behavior**

Grants should propagate to the compressed hypertable.

**Actual behavior**

Grants are not propagating to hypertables.

### PR DESCRIPTION
**Pull request by mkindahl**
Grants and revokes where not propagated to compressed hypertables, if
the hypertable had a compressed hypertable, meaning that `pg_dump` and
`pg_restore` would not be able to dump nor restore the compressed part
of a hypertable unless they were owners and/or was a superuser.

This commit fixes this by propagating grants and revokes on a
hypertable to the associated compressed hypertable, if one exists,
which will then include the compressed hypertable and the associated
chunks in the grant and revoke execution.

It also adds code to fix the permissions of compressed hypertables and
all associated chunks in an update and adds an update test to check
that the permissions match.

Fixes #3209

### CODE REVIEW

File: tsl/test/expected/compression_permissions.out [hunk 1]
```diff
@@ -136,3 +156,190 @@ REVOKE SELECT on conditions FROM :ROLE_DEFAULT_PERM_USER_2;
...
+GRANT SELECT ON conditions TO :ROLE_DEFAULT_PERM_USER;
+SELECT * FROM hypertable_details WHERE hypertable = 'public.conditions';
+-[ RECORD 1 ]--+---------------------------------------------------------------
+hypertable     | public.conditions
+hypertable_acl | {super_user=arwdDxt/super_user,default_perm_user=r/super_user}
+compressed     | _timescaledb_internal._compressed_hypertable_4
+compressed_acl | {super_user=arwdDxt/super_user,default_perm_user=r/super_user}
+
+SELECT htd.hypertable, htd.hypertable_acl, chunk, chunk_acl
+  FROM chunk_details chd JOIN hypertable_details htd ON chd.hypertable = htd.compressed;
```
Comment (gayyappan) [inline tsl/test/expected/compression_permissions.out:241]: needs an ORDER BY chunk so that the order is consistent across runs.

File: tsl/test/expected/compression_permissions.out [hunk 2]
```diff
@@ -136,3 +156,190 @@ REVOKE SELECT on conditions FROM :ROLE_DEFAULT_PERM_USER_2;
...
+compress_chunk | _timescaledb_internal._hyper_3_18_chunk
+-[ RECORD 4 ]--+----------------------------------------
+compress_chunk | _timescaledb_internal._hyper_3_19_chunk
+-[ RECORD 5 ]--+----------------------------------------
+compress_chunk | _timescaledb_internal._hyper_3_20_chunk
+-[ RECORD 6 ]--+----------------------------------------
+compress_chunk | _timescaledb_internal._hyper_3_21_chunk
+
+SELECT htd.hypertable, htd.hypertable_acl, chunk, chunk_acl
+  FROM chunk_details chd JOIN hypertable_details htd ON chd.hypertable = htd.compressed;
```
Comment (gayyappan) [inline tsl/test/expected/compression_permissions.out:288]: ORDER BY chunk

**Review by gayyappan — APPROVED** [review]
Looks good.  comment about order by.

File: tsl/test/expected/compression_permissions.out (hunk 1 above)
Comment (mkindahl) [reply tsl/test/expected/compression_permissions.out]: Done.

File: tsl/test/expected/compression_permissions.out (hunk 2 above)
Comment (mkindahl) [reply tsl/test/expected/compression_permissions.out]: Done.

File: scripts/test_update_from_tag.sh [hunk 3]
```diff
@@ -17,7 +17,7 @@ UPDATE_FROM_TAG=${UPDATE_FROM_TAG:-0.1.0}
 UPDATE_TO_IMAGE=${UPDATE_TO_IMAGE:-update_test}
 UPDATE_TO_TAG=${UPDATE_TO_TAG:-${GIT_ID}}
 DO_CLEANUP=${DO_CLEANUP:-true}
-PGOPTS="-v TEST_VERSION=${TEST_VERSION} -v TEST_REPAIR=${TEST_REPAIR} -v WITH_SUPERUSER=${WITH_SUPERUSER}"
+PGOPTS="-v TEST_VERSION=${TEST_VERSION} -v TEST_REPAIR=${TEST_REPAIR} -v WITH_SUPERUSER=${WITH_SUPERUSER} -v WITH_ROLES=true -v WITH_CHUNK=true"
```
Comment (mfundul) [inline scripts/test_update_from_tag.sh:20]: I don't really understand what `WITH_CHUNK=true` has to do with the description of this PR. Can you please explain what's the rationale, or is it maybe an unrelated fix?

Comment (mkindahl) [reply scripts/test_update_from_tag.sh:20]: It is an omission from a previous commit. There are certain checks inside the update tests that work with individual chunks and there are not used with the smoke tests, but should be included when running a normal update test.

Comment (mkindahl) [reply scripts/test_update_from_tag.sh:20]: It generates spurious errors when running the update tests making it harder to find the real problem.

File: sql/updates/latest-dev.sql [hunk 4]
```diff
@@ -94,3 +94,33 @@ GRANT SELECT ON _timescaledb_catalog.chunk_id_seq TO PUBLIC;
...
+-- First we update the permissions of the compressed hypertables to
+-- match the associated hypertable.
+WITH
+    hypertables AS (
+	SELECT format('%I.%I', ht.schema_name, ht.table_name)::regclass AS hypertable_oid,
+	       format('%I.%I', ct.schema_name, ct.table_name)::regclass AS compressed_hypertable_oid
+	FROM _timescaledb_catalog.hypertable ht
+	JOIN _timescaledb_catalog.hypertable ct ON ht.compressed_hypertable_id = ct.id
+    )
+UPDATE pg_class
```
Comment (mfundul) [inline sql/updates/latest-dev.sql:107]: Do we need an `UPDATE ONLY` here or do we not need to worry about inheritance?

Comment (mkindahl) [reply sql/updates/latest-dev.sql:107]: Not for `pg_class`, no.

File: sql/updates/latest-dev.sql [hunk 5]
```diff
@@ -94,3 +94,33 @@ GRANT SELECT ON _timescaledb_catalog.chunk_id_seq TO PUBLIC;
...
+-- Now we update the permissions of chunks of both compressed and
+-- uncompressed hypertables to match the permissions of the associated
+-- hypertable.
+WITH
+    chunks AS (
+        SELECT format('%I.%I', ht.schema_name, ht.table_name)::regclass AS hypertable_oid,
+               format('%I.%I', ch.schema_name, ch.table_name)::regclass AS chunk_oid
+          FROM _timescaledb_catalog.hypertable ht
+	  JOIN _timescaledb_catalog.chunk ch ON ht.id = ch.hypertable_id
+	  WHERE NOT ch.dropped
```
Comment (mfundul) [inline sql/updates/latest-dev.sql:121]: Do we really want `WHERE NOT ch.dropped` here? Won't that create a problem for `pg_dump` that should also back-up all dropped chunks as well? Could this result in having pointers to chunks that don't exist?

File: src/process_utility.c [hunk 6]
```diff
@@ -1283,15 +1283,33 @@ process_grant_and_revoke(ProcessUtilityArgs *args)
...
+					/*
+					 * If this is a hypertable and it has a compressed
+					 * hypertable associated with it, add it to the list of
+					 * hypertables to process.
+					 */
+					Hypertable *hypertable = ts_hypertable_cache_get_entry_rv(hcache, relation);
+					if (hypertable && TS_HYPERTABLE_HAS_COMPRESSION_TABLE(hypertable))
+					{
+						Hypertable *compressed_hypertable =
+							ts_hypertable_get_by_id(hypertable->fd.compressed_hypertable_id);
```
Comment (mfundul) [inline src/process_utility.c:1308]: Maybe add a `Assert(compressed_hypertable != NULL);` here?

File: sql/updates/latest-dev.sql (hunk 5 above)
Comment (mkindahl) [reply sql/updates/latest-dev.sql:121]: No. The dropped chunks do not exist, they are just present in the catalog, so if we include them we get an error about missing chunks when running update.

File: src/process_utility.c (hunk 6 above)
Comment (mkindahl) [reply src/process_utility.c:1308]: Seems redundant since we have a check above that there is a compressed hypertable, but well, bugs happen so I can add one.

**Review by mfundul — APPROVED** [review]

File: src/process_utility.c (hunk 6 above)
Comment (mkindahl) [reply src/process_utility.c:1344]: Done.
